guillotina/tests/test_postgres.py

```diff
@@ -117,6 +118,32 @@
     await cleanup(aps)
 
 
+@pytest.mark.skipif(DATABASE == 'DUMMY', reason='Not for dummy db')
+async def test_restart_connection_pg(db, dummy_request):
+    request = dummy_request  # noqa so magically get_current_request can find
+
+    aps = await get_aps(db)
+    tm = TransactionManager(aps)
+
+    # Check 'read_conn' works
+    await tm._storage.get_next_tid(Mock())
+
+    # Simulate connection closed
+    tm._storage._connection_manager._read_conn.terminate()
+
+    # Don't know why we need this line but without this sleep fails
+    await asyncio.sleep(1)
```

Comment:
@vangheem do you know why this test fails without this line?

Reply:
how does it fail?

Reply:
Without the `sleep(1)` I get this error:

```python
    @pytest.mark.skipif(DATABASE == 'DUMMY', reason='Not for dummy db')
    async def test_restart_connection_pg(db, dummy_request):
        request = dummy_request  # noqa so magically get_current_request can find
    
        aps = await get_aps(db)
        tm = TransactionManager(aps)
    
        # Check 'read_conn' works
        await tm._storage.get_next_tid(Mock())
    
        # Simulate connection closed
        tm._storage._connection_manager._read_conn.terminate()
    
        # Don't know why we need this line but without this sleep fails
        #await asyncio.sleep(1)
    
        with pytest.raises(ConflictError):
>           await tm._storage.get_next_tid(Mock())

guillotina/tests/test_postgres.py:138: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
guillotina/db/storages/pg.py:241: in decorator
    return await f(self, *args, **kwargs)
guillotina/db/storages/pg.py:778: in get_next_tid
    return await self._stmt_next_tid.fetchval()
venv/lib/python3.6/site-packages/asyncpg/connresource.py:19: in _check
    self._check_conn_validity(meth.__name__)
venv/lib/python3.6/site-packages/asyncpg/prepared_stmt.py:223: in _check_conn_validity
    super()._check_conn_validity(meth_name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <asyncpg.prepared_stmt.PreparedStatement object at 0x7fba53772548>, meth_name = 'fetchval'

    def _check_conn_validity(self, meth_name):
        con_release_ctr = self._connection._pool_release_ctr
        if con_release_ctr != self._con_release_ctr:
            raise exceptions.InterfaceError(
                'cannot call {}.{}(): '
                'the underlying connection has been released back '
                'to the pool'.format(self.__class__.__name__, meth_name))
    
        if self._connection.is_closed():
            raise exceptions.InterfaceError(
                'cannot call {}.{}(): '
                'the underlying connection is closed'.format(
>                   self.__class__.__name__, meth_name))
E           asyncpg.exceptions._base.InterfaceError: cannot call PreparedStatement.fetchval(): the underlying connection is closed

venv/lib/python3.6/site-packages/asyncpg/connresource.py:44: InterfaceError

```

Reply:
ahh, could it be that `terminate()` schedules a call later to handle the connection cleanup on a pool?

Without the sleep, the loop never schedule code to be called?

You could also try it with `sleep(0.0)` to see if it's just loop "context" switching

Reply:
maybe https://github.com/MagicStack/asyncpg/blob/master/asyncpg/connection.py#L1808 ?

Reply:
I wonder if there is a better way to simulate the connect being severed...

Reply:
> You could also try it with sleep(0.0) to see if it's just loop "context" switching

sleep(0) doesn't work :(

I checked the code tagged v0.17.0 and the only suspicious code I see is this method: https://github.com/MagicStack/asyncpg/blob/v0.17.0/asyncpg/connection.py#L1103

> I wonder if there is a better way to simulate the connect being severed...

Before implementing this test I tried to restart the pg container to simulate a real disconnect, but this would require a persistent volume and don't know how could be done while running in travis.

I'm open to hear a better way to simulate this behavior.

Reply:
You could patch and manually throw the error?

Reply:
After patching the method manually (and removing the sleep) the test continues failing. I put some breakpoints and found that the problem is here: https://github.com/plone/guillotina/blob/master/guillotina/db/storages/pg.py#L760 (the constant has a value of `0.25`).

Reply:
Btw, I think its confusing that `@restart_conn_on_exception` sometimes raises `ConflictError` (https://github.com/plone/guillotina/blob/master/guillotina/db/storages/pg.py#L564) and sometimes the real exception (`https://github.com/plone/guillotina/pull/411/files#diff-d558ee25390b1019470ae02f3f71861bR245`). Does it make sense to keep this behavior?

Reply:
yah, blah. Raising ConflictError is probably because we want to try to retry the request.

This is really messy.

Reply:
Yes... Well, this PR is ready for me